File: trunk/examples/echo/ControlBroker.cc

#include <iostream>
#include <fstream>
#include <ControlBroker.hh>
#include <ControlBroker.hpp>


ControlBroker::ControlBroker(CORBA::ORB_ptr orb, char *name) {
	m_orb = orb;
	m_name = name;
	m_nameServer = EDCBA::BrokerNameService::_nil();
}

ControlBroker::~ControlBroker() {
	//deregister();
}

// Registers with the BrokerNameService
bool ControlBroker::nsregister() {
	std::ifstream file;
	char buffer[500];
	
	// Find the name service
	file.open("/tmp/BrokerNameService.ior");
	file >> (char*)buffer;
	
	CORBA::Object_var obj = m_orb->string_to_object(buffer);
	m_nameServer = EDCBA::BrokerNameService::_narrow(obj);
	if( CORBA::is_nil(m_nameServer) ) {
		std::cerr << "Can't narrow reference to type BrokerNameService (or it was nil)." << std::endl;
		return 1;
	}
	
	// Register with the name service
	obj = _this();
  CORBA::String_var sior(m_orb->object_to_string(obj));
	//CORBA::String_var name((const char *)m_name.c_str());
	std::string auth = m_nameServer->nsregister(m_name.c_str(), m_orb->object_to_string(obj));
		
	if (auth != "") {
		m_auth = auth;
		return true;
	}
	m_nameServer = EDCBA::BrokerNameService::_nil();
	return false;
}
			
bool ControlBroker::deregister() {
	if (m_auth != "") {
		if (m_nameServer->deregister(m_auth.c_str(),m_name.c_str())) {
			m_auth = "";
			m_nameServer = EDCBA::BrokerNameService::_nil();
			return true;
		} else {
			std::cerr << "Unable to deregister" << std::endl;
			return false;
		}
	}
	return false;
}
			
bool ControlBroker::isRegistered() {
	return m_auth != "";
}
		
char* ControlBroker::getName() {
	return (char*)m_name.c_str();
}
		
void ControlBroker::serialize(const char *auth, const char *profile) {
}
	
void ControlBroker::deserialize(const char *auth, const char *profile) {
}

File: trunk/EdcbaBroker/ControlBroker.py
#!/usr/bin/env python

# The basic ControlBroker from which all others flow.
# Copyright 2006 John T. Kamenik, LGPL, All rights reserved.

import os, sys

try:
	from omniORB import CORBA
	import omniORB

	# for now assume we are being run from edcba root
	omniORB.importIDL("./idl/ControlBroker.idl")
	omniORB.importIDL("./idl/BrokerNameService.idl")

	import EDCBA__POA as EDCBA

	base = EDCBA.ControlBroker
except:
	# if we can't be a CORBA object, then we can't be anything
	base = object


class ControlBroker(base):
	def __init__(self, orb, name="Control Broker"):
		'''Creates a generic Control Broker'''
		self.name = name
		self.orb  = orb
		self.auth = None
		self.register()

	def __del__(self):
		'''Called when this class is cleaned up'''
		print "Deleting %s" % (self.name)
		try:
			if self.auth:
				ns.deregister(self.auth,self.name)
		except: pass

	def register(self):
		'''Registers with the BrokerNameService'''
		ior = file('/tmp/BrokerNameService.ior').read()
		obj = self.orb.string_to_object(ior)
		ns  = obj._narrow(EDCBA.BrokerNameService)
		auth = ns.nsregister(self.name,self.orb.object_to_string(self._this()))
		if auth:
			self.auth = auth
			self.ns   = ns
			return True
		else:
			return False

	def deregister(self):
		if self.auth is not None:
			if self.ns.deregister(self.auth,self.name):
				self.auth = None
				self.ns   = None
				return True
			else:
				print "Unable to deregister"
				return False
		else:
			return False

	def isRegistered(self):
		return self.auth is not None

	def getName(self):
		return self.name

	def serialize(self, auth, profile):
		pass

	def deserialize(self, auth, profile):
		pass


if __name__ == "__main__":
	orb = CORBA.ORB_init(sys.argv)

	obj = ControlBroker(orb)
	obj.deregister()
